### PR TITLE
Remove modules infrastructure and UI from the plugin

### LINKED ISSUE #1030: Remove modules infrastructure
Following #1029, there are effectively no longer any modules as part of Performance Lab. This should remain that way. As such, all modules infrastructure should be entirely removed.

Let's focus on the minimum effort needed to clean up in this issue. Further enhancements to user experience should be considered in separate issues.

Here is a non comprehensive list of what needs to be removed:
* All UI related to toggling modules (i.e. the checkboxes).
* All UI related to the module migration.
* All PHP functions related to module settings.
* All PHP functions related to loading modules, module activation/deactivation etc.
* All PHP functions related to migrating from modules to their standalone plugins.
* The `perflab_is_standalone_plugin_loaded()` PHP function (as it relates to modules).
* All unit test coverage related to module infrastructure.
* All CLI scripts in `bin` that deal with modules exclusively.
* All GitHub workflows that deal with modules exclusively (such as building and testing modules, publishing modules, ...).
* Utility files, like `module-i18n.php`, `default-enabled-modules.php`, and their usage and surrounding infrastructure.
* Any reference to modules, really :)

Additional changes to be made:
* The `plugins.json` file should no longer support a `"modules"` entry. Tooling (e.g. scripts and GitHub workflows) should be updated to no longer consider that `"modules"` entry. For example, the `deploy-standalone-plugins.yml` workflow should be updated to only work with the actual plugins in the `/plugins` folder and no longer reference modules, which means for example scripts like `get-plugin-dir` can be removed as it will now always be the `/plugins` directory.
* The `$source` parameter of the `perflab_get_standalone_plugin_version_constants()` function should be deprecated (or removed) and effectively always be for plugins.
* The generator tag should be adjusted to not include any information about modules. In other words, it output should look like `Performance Lab %1$s; plugins: %2$s` going forward.
* The copy of the default admin pointer should be adjusted to no longer say "performance features included in the plugin", but just "performance features".
* The main `readme.txt` should be updated to no longer list the available modules.
* The `readme` JS script should be adjusted to no longer try to populate the list of available modules into `readme.txt` (the changelog functionality should be kept though).

These lists may not be complete, and further things to remove/update may come up as a PR is being worked on.

### PR DESCRIPTION
**Pull request by mukeshpanchal27**
## Summary

Fixes #1030

Things needs to updates:
- Document: https://make.wordpress.org/performance/handbook/performance-lab/releasing-the-plugin/

**Note:** In `.github/workflows/deploy-standalone-plugins.yml`, I've included some 'TODO' items that need to be addressed and removed before the merge.

### CODE REVIEW

**Comment by mukeshpanchal27** [comment]
For the CODEOWNERS files:
- Should we remove `Focus` areas?
- Do we need to add a new section called `Site Health`, and include `WebP Support Health Check`, `Autoloaded Options Health Check`, and `Enqueued Assets Health Check` with their respective owners? What do you think?

File: .github/workflows/deploy-standalone-plugins.yml
Comment (thelovekesh) [inline .github/workflows/deploy-standalone-plugins.yml:1]: This workflow is updated at https://github.com/WordPress/performance/pull/1033. @felixarntz @mukeshpanchal27 Can we please aim to merge https://github.com/WordPress/performance/pull/1033 before we merge this PR?

File: plugins.json
Comment (thelovekesh) [inline plugins.json:1]: Should this be updated to
```json
[
  "auto-sizes",
  "dominant-color-images",
  "speculation-rules",
  "webp-uploads"
]
```
Since `plugins.json` only store plugin names anyway.

File: readme.txt
Comment (thelovekesh) [inline readme.txt]: > Existing plugins may be removed after they have been released in WordPress core, while new modules may be added in any release.

I think modules is confusing here given we are removing them completely. Should we term it as standalone or feature plugin here?

File: admin/load.php
Comment (thelovekesh) [inline admin/load.php]: Do we need to change the functions name as well?

File: .github/workflows/deploy-standalone-plugins.yml
Comment (felixarntz) [reply .github/workflows/deploy-standalone-plugins.yml:1]: @thelovekesh That works for me. @mukeshpanchal27 What do you think? If we merge #1033 first, there will be merge conflicts here in the workflow files. But since #1033 makes more complex changes to those workflows, it's probably easier to merge that and then fix the little extra things in here.

File: admin/load.php
Comment (felixarntz) [reply admin/load.php]: +1

Maybe just call it `perflab_add_settings_page()`, so it's not tied to modules/plugins internals. Or it could be called `perflab_add_features_page()`. WDYT?

You can add `@since` annotations like `@since n.e.x.t Renamed to ...`.

File: admin/load.php
Comment (felixarntz) [inline admin/load.php]: How about we already name this "Performance Features" instead, in preparation for #1031?

File: admin/load.php
Comment (felixarntz) [inline admin/load.php]: Why this change? This adds those action hooks unconditionally, which means they may load all over WP Admin.

I think we should maintain `perflab_load_modules_page()` (just renamed to `perflab_load_settings_page()`, similar to @thelovekesh's previous suggestion) and keep the action hooks in there.

File: admin/load.php
Comment (felixarntz) [inline admin/load.php]: See above, can you rename this accordingly as well?

File: plugins.json
Comment (felixarntz) [reply plugins.json:1]: Maybe. I'd say let's keep as is for now as the extra property doesn't really hurt, and keeping it avoids having to make more changes in this PR.

File: readme.txt
Comment (felixarntz) [inline readme.txt]: See #1031: We should focus on the _features_ rather than that they are plugins.
```suggestion
The Performance Lab plugin is a collection of features focused on enhancing performance of your site, most of which should eventually be merged into WordPress core. The plugin allows to individually enable and test the features to get their benefits before they become available in WordPress core, and to provide feedback to further improve the solutions.
```

File: readme.txt
Comment (felixarntz) [inline readme.txt]:
```suggestion
2. Enable the individual features you would like to use.
```

File: readme.txt
Comment (felixarntz) [inline readme.txt]:
```suggestion
The primary purpose of the Performance Lab plugin is to allow testing of various performance features for which the goal is to eventually land in WordPress core. It is essentially a collection of "feature plugins", which makes it different from other performance plugins that offer performance features which are not targeted at WordPress core and potentially rely on functionality that would not be feasible to use in WordPress core. The list of available features will regularly change: Existing features may be removed after they have been released in WordPress core, while new features may be added in any release.
```

File: load.php
Comment (felixarntz) [inline load.php]:
```suggestion
 * @since n.e.x.t The generator tag no longer includes module slugs.
```

File: load.php
Comment (felixarntz) [inline load.php]:
```suggestion
 * @since n.e.x.t The $source parameter was removed.
```

**Review by felixarntz — COMMENTED** [review]
@mukeshpanchal27 This looks great so far. I left a few comments.

File: .github/workflows/deploy-standalone-plugins.yml
Comment (mukeshpanchal27) [reply .github/workflows/deploy-standalone-plugins.yml:1]: Yes, I will take care of that. Thank you for the heads up.

File: admin/load.php
Comment (mukeshpanchal27) [reply admin/load.php]: Addressed in https://github.com/WordPress/performance/pull/1060/commits/66f0492278e40c1c5a31790b9d20515977e6d3b3

**Comment by thelovekesh** [comment]
@mukeshpanchal27 #1033 caused merge conflicts 😐

File: .github/workflows/deploy-standalone-plugins.yml
Comment (felixarntz) [inline .github/workflows/deploy-standalone-plugins.yml]: Please remember to revert this once testing is complete.

Also see the merge conflicts with this file based on #1033.

File: admin/load.php
Comment (felixarntz) [inline admin/load.php]: Since you went with `features_page` in the other function above, please align that here as well.

File: readme.txt
Comment (felixarntz) [inline readme.txt]: Please update the plugin description like this everywhere, including in `load.php`, `composer.json` etc.
```suggestion
Performance plugin from the WordPress Performance Team, which is a collection of standalone performance features.
```

**Comment by mukeshpanchal27** [comment]
@felixarntz, should we update the setting page slug from the existing `/wp-admin/options-general.php?page=perflab-modules` to `/wp-admin/options-general.php?page=perflab-features`?

**Comment by thelovekesh** [comment]
Overall changes look good to me. I just want to verify whether an updater script is required to remove any options or configurations added to the database by the previous PL plugin that will not be utilized in the next version.

**Comment by felixarntz** [comment]
@mukeshpanchal27 
> @felixarntz, should we update the setting page slug from the existing `/wp-admin/options-general.php?page=perflab-modules` to `/wp-admin/options-general.php?page=perflab-features`?

Great catch! I think this is a good idea, though I would suggest we go with the more generally applicable `/wp-admin/options-general.php?page=performance-lab`. Let's also rename the constant, e.g. to `PERFLAB_SCREEN` or `PERFLAB_FEATURES_SCREEN`.

It would be great to add another hook though which catches requests to `/wp-admin/options-general.php?page=perflab-modules` and redirects those to the new location, e.g. on the `admin_page_access_denied` action hook. For reference, consider [this example](https://github.com/google/site-kit-wp/blob/12bb2a84fa62f9b329ffd750be6876001bc6dd11/includes/Core/Admin/Screens.php#L317) in the Google Site Kit plugin.

@thelovekesh
> I just want to verify whether an updater script is required to remove any options or configurations added to the database by the previous PL plugin that will not be utilized in the next version.

Great point as well! It would be great to implement an update routine which deletes the `'perflab_modules_settings'` option if it is present. Since the option would be autoloaded anyway, I think we could maybe do that simply on `admin_init` if the current user has `manage_options` capabilities. This check should be cheap enough to run in that action hook, and it has the benefit that it works reliably at scale, even on large multisites.

File: .github/workflows/php-test-standalone-plugins.yml
Comment (westonruter) [inline .github/workflows/php-test-standalone-plugins.yml:80]: Why are the plugins no longer needing to be built? In `bin/plugin/commands/test-plugins.js` it is not accounting for assets which may have been generated by Webpack as it is just copying the source files from `/plugins` to `/build`. It seems rather that `test-plugins.js` should be updated to reuse the output from `npm run build-plugins` which is already outputting the plugins into `/build`. cc @thelovekesh

File: bin/plugin/commands/test-plugins.js
Comment (westonruter) [inline bin/plugin/commands/test-plugins.js:436]: Per above, is this entire try/catch block even needed if we already have `npm run build-plugins`?

File: uninstall.php
Comment (westonruter) [inline uninstall.php:40]: Should this remain? What if someone was on Performance Lab v2 and then they upgrade to v3 but then uninstall. Won't this option be then remaining? Perhaps leave this line here but add a comment to say it was from an older version. 

That brings to mind that it seems at present there is an _autoloaded_ `perflab_modules_settings` option which is getting loaded needlessly. It seems there should be an upgrade routine added for 2.9.0 to 3.0.0 which deletes this option or at least makes it not be autoloaded.

**Review by westonruter — CHANGES_REQUESTED** [review]

File: uninstall.php
Comment (westonruter) [reply uninstall.php:40]: Oh, this may be what @thelovekesh mentioned in https://github.com/WordPress/performance/pull/1060#issuecomment-2009992383.

Comment (westonruter) [reply uninstall.php:40]: Heh, and @felixarntz in https://github.com/WordPress/performance/pull/1060#issuecomment-2010175828.

Comment (felixarntz) [reply uninstall.php:40]: @westonruter Yeah, I don't think we need to keep it if we instead have a "migration" logic that deletes this option after an update. That way the old option is only referenced in one place, and it makes more sense to delete it after the update rather than only once (or rather if) the plugin is uninstalled.

File: .github/workflows/php-test-standalone-plugins.yml
Comment (mukeshpanchal27) [reply .github/workflows/php-test-standalone-plugins.yml:80]: Thank you for pointing that out. It's on my list to address as a follow-up PR since it scratches the current one.

**Comment by mukeshpanchal27** [comment]
@westonruter, I'll update the `test-plugins` command in a follow-up PR.

@felixarntz @thelovekesh, the feedback has been addressed. The PR is now ready for another round of review.

File: load.php
Comment (thelovekesh) [inline load.php]: Do we need the `get_option()` check? I think `delete_option()` handles it on its own.

**Review by thelovekesh — APPROVED** [review]
LGTM! Just a minor feedback ;)

File: bin/plugin/commands/test-plugins.js
Comment (thelovekesh) [inline bin/plugin/commands/test-plugins.js:1]: Do we still need this given we aren't converting modules to plugins anymore? I am not sure if we really need this.

File: load.php
Comment (mukeshpanchal27) [reply load.php]: Remove it in https://github.com/WordPress/performance/pull/1060/commits/959377cec9dddc5d8997b7bb5e9bd03e49868f90

File: bin/plugin/commands/test-plugins.js
Comment (mukeshpanchal27) [reply bin/plugin/commands/test-plugins.js:1]: That command simply checks if the plugin activates and doesn't return any errors, so I don't believe we need to remove it. I'd appreciate getting other opinions on this matter.

File: .github/workflows/php-test-standalone-plugins.yml
Comment (felixarntz) [reply .github/workflows/php-test-standalone-plugins.yml:80]: Based on what @thelovekesh wrote in https://github.com/WordPress/performance/pull/1060#discussion_r1534051355, do we even need this anymore? The purpose of the `test-plugins` script and this GitHub workflow was purely to test whether standalone plugins don't cause some weird fatal error when built from their modules.

Since we don't build modules into plugins anymore, I think this isn't useful any more. For that reason, I would suggest to remove the whole `test-plugins` script and everything around it. Since this is only removing more things, it would be great if you could do it as part of this PR.

I think some of the code changes in this PR are a bit confusing because they're changing those scripts when they don't really have a good purpose anymore as a whole.

**Review by felixarntz — COMMENTED** [review]
@mukeshpanchal27 Unless I'm missing something, I think we should remove the `test-plugins` script and everything using it, including the workflow, as this was intended to test plugins that were built from modules, which is no longer relevant. See https://github.com/WordPress/performance/pull/1060/files#r1534169130.

cc @westonruter @thelovekesh

File: bin/plugin/commands/test-plugins.js
Comment (felixarntz) [reply bin/plugin/commands/test-plugins.js:1]: Just replied in https://github.com/WordPress/performance/pull/1060#discussion_r1534169130 :)

**Comment by westonruter** [comment]
> Unless I'm missing something, I think we should remove the `test-plugins` script and everything using it, including the workflow, as this was intended to test plugins that were built from modules, which is no longer relevant. See https://github.com/WordPress/performance/pull/1060/files#r1534169130.

I think there is a better way to check for fatals and that would be through E2E tests in https://github.com/WordPress/performance/issues/1051. I think it makes sense to remove the existing `test-plugins` script as it currently stands.

This can be done in a separate PR, AFAIK.

File: load.php
Comment (westonruter) [inline load.php]: This would be better (as in more testable) if the `exit` only happens conditionally based on whether the redirect happens:

```suggestion
	if ( 
		current_user_can( 'manage_options' ) && 
		wp_safe_redirect( add_query_arg( 'page', PERFLAB_SCREEN ) ) 
	) {
		exit;
	}
```

But apparently this isn't being tested right now, so not critical to change here.

**Review by westonruter — APPROVED** [review]
Just a nit, but this looks like it can be merged and then further iterated on in subsequent PRs.

**Review by felixarntz — APPROVED** [review]
I agree with @westonruter's assessment, let's merge this to unblock other PRs. We can remove the `test-plugins` command in a subsequent PR.

**Pull request merged**
